mypy/checkexpr.py

```diff
@@ -1096,18 +1097,14 @@ def check_boolean_op(self, e: OpExpr, context: Context) -> Type:
             right_map, left_map = \
                 mypy.checker.find_isinstance_check(e.left, self.chk.type_map,
                                                    self.chk.typing_mode_weak())
+            truthiness_restriction = false_only
         elif e.op == 'or':
             left_map, right_map = \
                 mypy.checker.find_isinstance_check(e.left, self.chk.type_map,
                                                    self.chk.typing_mode_weak())
-        else:
```

Comment:
I'm adding an assertion about e.op at the beginning (which is assured by `visit_op_expr`)